src/System.Management.Automation/engine/Modules/ModuleSpecification.cs

```diff
@@ -266,92 +273,60 @@ public static bool TryParse(string input, out ModuleSpecification result)
-                {
-                    if (x.Guid.HasValue && y.Guid.HasValue)
-                    {
-                        result = x.Guid.Equals(y.Guid);
-                    }
-                }
-                if (result)
-                {
-                    if (x.Version != null && y.Version != null)
-                    {
-                        result = x.Version.Equals(y.Version);
-                    }
-                    else if (x.Version != null || y.Version != null)
-                    {
-                        result = false;
-                    }
-
-                    if (x.MaximumVersion != null && y.MaximumVersion != null)
-                    {
-                        result = x.MaximumVersion.Equals(y.MaximumVersion);
-                    }
-                    else if (x.MaximumVersion != null || y.MaximumVersion != null)
-                    {
-                        result = false;
-                    }
-
-                    if (result && x.RequiredVersion != null && y.RequiredVersion != null)
-                    {
-                        result = x.RequiredVersion.Equals(y.RequiredVersion);
-                    }
-                    else if (result && (x.RequiredVersion != null || y.RequiredVersion != null))
-                    {
-                        result = false;
-                    }
-                }
+                return true;
             }
 
-            return result;
+            return x != null && y != null
+                && String.Equals(x.Name, y.Name, StringComparison.OrdinalIgnoreCase)
+                && Guid.Equals(x.Guid, y.Guid)
+                && Version.Equals(x.RequiredVersion, y.RequiredVersion)
+                && Version.Equals(x.Version, y.Version)
+                && String.Equals(x.MaximumVersion, y.MaximumVersion);
         }
 
+        /// <summary>
+        /// Get a property-based hashcode for a ModuleSpecification object.
+        /// </summary>
+        /// <param name="obj">The module specification for the object.</param>
+        /// <returns>A hashcode that is always the same for any module specification with the same properties.</returns>
+        /// <remarks>
+        /// The algorithm in this method is adapted from a StackOverflow answer from Jon Skeet: https://stackoverflow.com/a/263416/9944203
+        /// </remarks>
         public int GetHashCode(ModuleSpecification obj)
         {
-            int result = 0;
+            if (obj == null)
+            {
+                return 0;
+            }
+
+            // Initial prime value
+            int p = 23;
+
+            // Use 31 for multiplication. See https://computinglife.wordpress.com/2008/11/20/why-do-hash-functions-use-prime-numbers/
+            int q = 31;
+
+            int result = p; 
 
-            if (obj != null)
+            foreach (object property in new object[] { obj.Name, obj.Guid, obj.RequiredVersion, obj.Version, obj.MaximumVersion })
             {
```

Comment:
While the loop is nicer code it does mean that you doing an allocation where there wasn't one before.

Reply:
Yes I tried to look for a better way to do this (without the allocation), but I'm guessing it's not possible. Either we keep the allocation or we unroll the loop. Thought the loop was nicer initially, but can go either way.

Reply:
I agree that we should avoid such allocations.
As we discussed above we could review all code base to clean up GetHashCode methods to use unified approach.

Reply:
So everybody put the same questions I suggest now use 
`internal static int CombineHashCodes(int h1, int h2, int h3, int h4, int h5)`
and defer improvements to new PR.

Reply:
I imagine `CombineHashCodes` may not get improved for a while if ever and for large numbers of modules, a better distribution is helpful. How about if I unroll the loop but use the new hash algorithm?

Reply:
Ok.

Reply:
If we do end up updating the `CombineHashCodes` method, we should circle back here and call that.